src/Network/Pusher/Data.hs

```diff
@@ -201,3 +226,136 @@ type Event = T.Text
+
+instance A.FromJSON Interest where
+  parseJSON v =
+    case v of
+      A.String s ->
+        case mkInterest s of
+          Nothing ->
+            fail $
+            "An Interest contains invalid characters or is too long: " ++ show s
+          Just istr -> pure istr
+      _ -> failExpectStr v
+
+instance A.ToJSON Interest where
+  toJSON (Interest txt) = A.String txt
+
+-- | URL to which pusher will send information about sent push notifications
+type WebhookURL = T.Text
+
+-- | Level of detail sent to WebhookURL. Defaults to Info
+data WebhookLevel
+  = Info -- ^ Errors only
+  | Debug -- ^ Everything
+
+instance A.FromJSON WebhookLevel where
+  parseJSON v =
+    case v of
+      A.String s
+        | s == "INFO" -> pure Info
+        | s == "DEBUG" -> pure Debug
+      _ -> failExpectStr v
+
+instance A.ToJSON WebhookLevel where
+  toJSON w =
+    A.String $
+    case w of
+      Info -> "INFO"
+      Debug -> "DEBUG"
+
+-- | Apple push notification service payload
+-- TODO: Replace JSON object with a stronger encoding.
+data APNSPayload =
+  APNSPayload A.Object
+
+instance A.FromJSON APNSPayload where
+  parseJSON v =
+    case v of
+      A.Object o -> pure . APNSPayload $ o
+      _ -> failExpectObj v
+
+instance A.ToJSON APNSPayload where
+  toJSON (APNSPayload o) = A.Object o
+
+-- | Google Cloud Messaging payload
+-- TODO: Replace JSON object with a stronger encoding.
+data GCMPayload =
+  GCMPayload A.Object
+
+instance A.FromJSON GCMPayload where
+  parseJSON v =
+    case v of
+      A.Object o -> pure . GCMPayload $ o
+      _ -> failExpectObj v
+
+instance A.ToJSON GCMPayload where
+  toJSON (GCMPayload o) = A.Object o
+
+-- | Firebase Cloud Messaging payload
+-- TODO: Replace JSON object with a stronger encoding.
+data FCMPayload =
+  FCMPayload A.Object
+
+instance A.FromJSON FCMPayload where
+  parseJSON v =
+    case v of
+      A.Object o -> pure . FCMPayload $ o
+      _ -> failExpectObj v
+
+instance A.ToJSON FCMPayload where
+  toJSON (FCMPayload o) = A.Object o
+
+data Notification = Notification
+  { notificationInterest :: Interest
+  , notificationWebhookURL :: Maybe WebhookURL
+  , notificationWebhookLevel :: Maybe WebhookLevel
+  , notificationAPNSPayload :: Maybe APNSPayload
+  , notificationGCMPayload :: Maybe GCMPayload
+  , notificationFCMPayload :: Maybe FCMPayload
+  }
+
+instance A.FromJSON Notification where
+  parseJSON (A.Object v) =
+    Notification <$>
+    (do interests <- v A..: "interests"
+        case interests of
+          A.Array arr
+            | V.length arr == 1 -> A.parseJSON $ V.head arr
+            | otherwise -> failExpectSingletonArray interests
+          v' -> failExpectArray v') <*>
+    v .:? "webhook_url" <*>
+    v .:? "webhook_level" <*>
+    v .:? "apns" <*>
+    v .:? "gcm" <*>
+    v .:? "fcm"
+  parseJSON v = failExpectObj v
+
+instance A.ToJSON Notification where
+  toJSON (Notification interests mWebhookURL mWebhookLevel mAPNS mGCMP mFCMP) =
```

Comment:
It's a convention I see people use sometimes to indicate the type is encapsulated in a 'Maybe'. I've only used it here as extra emphasis because if removed a person might be more temped to remove the following 'consOptionals' function and just use a toJSON call which would have the effect of creating null pairs (E.G. `fcm : null`) rather than omitting the pair all together - which is the behavior we want.
It might be better to leave this thought in a comment.